project/src/com/google/daggerquery/executor/models/Query.java

```diff
@@ -79,20 +86,71 @@ public Query(String typeName, String... parameters) {
   /**
    * Executes query on a {@link BindingGraph}.
    *
-   * <p>For all queries return a list with string. Strings content depends on a query name.
-   * <p>For `deps` queries each string represent exactly one dependency.
+   * <p>For all queries return a list with strings. Strings content depends on a query name.
+   *
+   * <ul>
+   * <li>For `deps` query each string represent exactly one dependency.
+   * <li>For `allpaths` query each string contains a path between {@code source} and {@code target} nodes.
+   * The connection between nodes is shown with construction '->'. For example, one of the possible paths
+   * may look like this: "com.google.First -> com.google.Second -> com.google.Third".
+   * </ul>
    */
   public List<String> execute(BindingGraph bindingGraph) {
-    if (name.equals("deps")) {
-      String sourceNode = parameters[0];
-      if (!bindingGraph.getAdjacencyListMap().containsKey(sourceNode)) {
-        throw new IllegalArgumentException("Specified source node doesn't exist.");
-      }
+    String source = parameters[0];
 
-      return bindingGraph.getAdjacencyListMap().get(sourceNode).getDependencyList()
-          .stream().map(DependencyProto.Dependency::getTarget).sorted().collect(toList());
+    if (!bindingGraph.getAdjacencyListMap().containsKey(source)) {
+      throw new IllegalArgumentException("Specified source node doesn't exist.");
+    }
+
+    switch (name) {
+      case DEPS_QUERY_NAME:
+        return bindingGraph.getAdjacencyListMap().get(source).getDependencyList()
+            .stream().map(Dependency::getTarget).sorted().collect(toList());
+
+      case ALLPATHS_QUERY_NAME:
+        String target = parameters[1];
+        Set<String> visitedNodes = new HashSet<>();
+        List<List<String>> result = new ArrayList<>();
+        List<String> path = new ArrayList<>();
+
+        findAllPaths(source, target, path, bindingGraph, visitedNodes, result);
```

Comment:
This may be more of a hack (which I have used myself), is to use Guava's [`ImmutableSet`](https://guava.dev/releases/19.0/api/docs/com/google/common/collect/ImmutableSet.html). Guava's [`ImmutableCollection`](https://github.com/google/guava/wiki/ImmutableCollectionsExplained#how) classes guarantee the iteration order to be the same as construction order. You may be able to cheekily use this for the `Path` implementation, if the `Path` never contains duplicate elements (cycles).

Reply:
Sorry I forgot that I proposed using `ArrayDeque` earlier. Yeah I'm not sure if there 's a data structure which implements both `Set` and `Stack`. Please ignore.